image_widget_crop.module

```diff
@@ -203,3 +204,236 @@ function image_widget_crop_library_info_alter(&$libraries, $extension) {
+
+    if ($old_properties['edit'] == TRUE) {
+      if ($new_properties['crop_applied'] == 1) {
+        if ($image_widget_crop_manager->cropHasChanged($new_properties, $old_properties)) {
+          $image_widget_crop_manager->updateCrop($new_properties, $old_properties, $crop_type[$crop_type_id]);
+        }
+      }
+      elseif ($new_properties['crop_applied'] == 0) {
+        $image_widget_crop_manager->deleteCrop($old_properties['file-uri'], $crop_type[$crop_type_id], $old_properties['file-id']);
+      }
+    }
+
+  }
+}
+
+/**
+ * Implements hook_form_alter().
+ */
+function image_widget_crop_form_alter(&$form, \Drupal\Core\Form\FormStateInterface $form_state, $form_id) {
+  // We only want crop ui on images.
+  if (substr($form_id, 0, 5) == 'file_' && (substr($form_id, -10, 10) == '_edit_form' || substr($form_id, -17, 17) == '_inline_edit_form')) {
+    $entity = $form_state->getFormObject()->getEntity();
+    $mime_type = $entity->getMimeTypeType();
+
+    if ($mime_type !== 'image') {
+      return;
+    }
+    $element['#theme'] = 'image_widget';
+    $element['#attached']['library'][] = 'image/form';
+    $form['#attached']['library'][] = 'image_widget_crop/cropper.integration';
+
+    $form['crop_preview_wrapper'] = [
+      '#type' => 'details',
+      '#title' => t('Crop image'),
+      '#attributes' => ['class' => ['crop-wrapper']],
+      '#weight' => 100,
+      '#tree' => TRUE,
+    ];
+
+    // Warn the user if the crop is used more than once.
+    $usage_counter = 0;
+    $file_usage = \Drupal::service('file.usage')->listUsage($entity);
+    foreach (new RecursiveIteratorIterator(new RecursiveArrayIterator($file_usage)) as $usage) {
+      $usage_counter += (int) $usage;
+    }
+    if ($usage_counter > 1) {
+      $form['crop_reuse'] = [
+        '#type' => 'container',
+        '#markup' => t('This crop definition affects more usages of this image'),
+        '#attributes' => [
+          'class' => ['messages messages--warning'],
+        ],
+        '#weight' => -10,
+      ];
+    }
+
+    $element_wrapper_name = 'crop_container';
+    // Ensure that the ID of an element is unique.
+    $list_id = \Drupal::service('uuid')->generate();
+
+    $container = &$form['crop_preview_wrapper'];
+    $container[$list_id] = [
+      '#type' => 'vertical_tabs',
+      '#default_tab' => '',
+      '#theme_wrappers' => array('vertical_tabs'),
+      '#tree' => TRUE,
+      '#parents' => array($list_id),
+    ];
+
+    $crop_type_list = \Drupal::service('entity.manager')
+      ->getStorage('crop_type')
+      ->loadMultiple();
+
+    if (!empty($crop_type_list)) {
+      foreach ($crop_type_list as $crop_type) {
+        $crop_type_id = $crop_type->id();
+        $label = $crop_type->label();
+
+        $container[$crop_type_id] = [
+          '#type' => 'details',
+          '#title' => $label,
+          '#group' => $list_id,
+        ];
+
+        $thumb_properties = [];
+
+        // Add compatibility to PHP 5.3.
```

Comment:
It would also be trivial to write this in a 5.3 compatible way like this: $crop_type->getAspectRatio() ?: t('NaN');